### PR TITLE
Search Queries in parallel - part 3

### PR DESCRIPTION
**Pull request by drempapis**
please see https://github.com/elastic/elasticsearch/pull/116812

### CODE REVIEW

**Comment by drempapis** [comment]
@elasticmachine update branch

**Comment by drempapis** [comment]
@elasticmachine update branch

File: server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 1]
```diff
@@ -519,25 +499,14 @@ public void testSearchingFilteringAliasesMultipleIndices() throws Exception {
...
-            prepareSearch("filter13", "filter1").setSize(0).setQuery(QueryBuilders.matchAllQuery()),
-            searchResponse -> assertThat(searchResponse.getHits().getTotalHits().value(), equalTo(4L))
-        );
-
         assertResponse(
             prepareSearch("filter13", "filter1", "filter23").setQuery(QueryBuilders.matchAllQuery()),
             searchResponse -> assertHits(searchResponse.getHits(), "11", "12", "13", "21", "31", "33")
         );
-        assertResponse(
-            prepareSearch("filter13", "filter1", "filter23").setSize(0).setQuery(QueryBuilders.matchAllQuery()),
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java:537]: I think that we may have accidentally removed this one completely?

File: server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java [hunk 2]
```diff
@@ -149,33 +150,24 @@ public void testRescorePhrase() throws Exception {
...
-                assertThat(response.getHits().getHits()[2].getId(), equalTo("2"));
+                assertFirstHit(response, hasId("1"));
+                assertSecondHit(response, hasId("3"));
+                assertThirdHit(response, hasId("2"));
             }
         );
-        assertResponse(
+        assertResponses(response -> {
+            assertHitCount(response, 3);
+            assertThat(response.getHits().getMaxScore(), equalTo(response.getHits().getHits()[0].getScore()));
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java:162]: I would assume that this is safe, but this is a new addition for this test block, right?

Comment (pmpailis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/functionscore/QueryRescorerIT.java:162]: Just noticed it below; in L#174 :)

File: server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java [hunk 3]
```diff
@@ -630,147 +623,87 @@ public void testCrossFieldMode() throws ExecutionException, InterruptedException
...
                         MultiMatchQueryBuilder.Type.CROSS_FIELDS
                     ).analyzer("category").operator(Operator.OR)
                 )
-            ),
-            response -> assertFirstHit(response, hasId("theone"))
+            )
         );
 
-        // test group based on analyzer -- all fields are grouped into a cross field search
-        assertResponse(
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java:761]: I think that this one is missing too from the refactoring?

File: server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java (hunk 1 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java:537]: Thank you, @pmpailis, for finding this. I'll add it back to the refactored code.

File: server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java (hunk 3 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java:761]: Thank you Pano!

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 4]
```diff
@@ -1135,15 +1124,11 @@ private void checkMatchedFieldsCase(boolean requireFieldMatch, String type) thro
...
             req.setQuery(queryStringQuery("running scissors").field("foo")),
-            response -> assertHighlight(response, 0, "foo", 0, equalTo("<em>running</em> with <em>scissors</em>"))
-        );
-
-        // Now make half the matches come from the stored field and half from just a matched field.
-        assertResponse(
-            req.setQuery(queryStringQuery("foo.plain:running scissors").field("foo")),
-            response -> assertHighlight(response, 0, "foo", 0, equalTo("<em>running</em> with <em>scissors</em>"))
+            // Now make half the matches come from the stored field and half from just a matched field.
+            req.setQuery(queryStringQuery("foo.plain:running scissors").field("foo"))
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1131]: Wouldn't this change the existing `req` instance ? So we would end up running the 2nd test-case twice?

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 5]
```diff
@@ -1161,50 +1146,35 @@ private void checkMatchedFieldsCase(boolean requireFieldMatch, String type) thro
...
-            req.setQuery(queryStringQuery("foo.plain:running^5 scissors").field("foo")),
-            response -> assertHighlight(response, 0, "foo", 0, equalTo("<em>running</em> with <em>scissors</em>"))
-        );
 
-        // Now just all matches are against the matched field. This still returns highlighting.
-        assertResponse(
+        assertResponses(
+            response -> assertHighlight(response, 0, "foo", 0, equalTo("<em>running</em> with <em>scissors</em>")),
+            // Now make sure boosted fields don't blow up when matched fields is both the subfield and stored field.
+            req.setQuery(queryStringQuery("foo.plain:running^5 scissors").field("foo")),
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1160]: Same as above; are we reusing the same `Request` instance in these tests?

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 6]
```diff
@@ -2352,22 +2288,14 @@ public void testHighlightNoMatchSizeNumberOfFragments() {
...
-            assertHighlight(response, 0, "text", 1, 2, equalTo("This is the <em>fifth</em> sentence"));
-        });
-        field.highlighterType("unified");
-        assertResponse(prepareSearch("test").setQuery(queryBuilder).highlighter(new HighlightBuilder().field(field)), response -> {
+        assertResponses(response -> {
             assertHighlight(response, 0, "text", 0, 2, equalTo("This is the <em>third</em> sentence. This is the fourth sentence."));
             assertHighlight(response, 0, "text", 1, 2, equalTo("This is the <em>fifth</em> sentence"));
-        });
+        },
+            prepareSearch("test").setQuery(queryBuilder).highlighter(new HighlightBuilder().field(field.highlighterType("plain"))),
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:2295]: Same as above, don't we update the same `field` used, so in practice we're only testing `unified`?

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 7]
```diff
@@ -1214,24 +1184,22 @@ private void checkMatchedFieldsCase(boolean requireFieldMatch, String type) thro
...
             .requireFieldMatch(requireFieldMatch);
-        assertResponse(req.setQuery(twoFieldsQuery).highlighter(new HighlightBuilder().field(fooField).field(barField)), response -> {
+        assertResponses(response -> {
             assertHighlight(response, 0, "foo", 0, equalTo(expectedHighlight1));
             assertHighlight(response, 0, "bar", 0, equalTo("<em>cat</em> <em>cat</em> junk junk junk junk"));
-        });
+        },
+            req.setQuery(twoFieldsQuery).highlighter(new HighlightBuilder().field(fooField).field(barField)),
+            // Setting a matchedField that isn't searched/doesn't exist is simply ignored.
+            req.setQuery(twoFieldsQuery).highlighter(new HighlightBuilder().field(fooField).field(barField.matchedFields("bar", "candy")))
```
Comment (pmpailis) [inline server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1193]: Same as above; reusing the same `req` instance?

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 4 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1131]: You are right!! I missed this spot. I'll figure it out. Thank you Pano

Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1131]: Fixed it

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 5 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1160]: Done!

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 7 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java:1193]: Done!

File: server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java (hunk 3 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/query/MultiMatchQueryIT.java:761]: Added back

File: server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 6 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Revert

**Review by pmpailis — APPROVED** [review]
LGTM! Thanks @drempapis !

File: server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java [hunk 8]
```diff
@@ -324,27 +316,15 @@ public void testSearchingFilteringAliasesSingleIndex() throws Exception {
...
-            searchResponse -> assertHits(searchResponse.getHits(), "1", "2", "3", "4")
-        );
-
         logger.info("--> checking index and filtering alias search");
-        assertResponse(
-            prepareSearch("test", "foos").setQuery(QueryBuilders.matchAllQuery()),
-            searchResponse -> assertHits(searchResponse.getHits(), "1", "2", "3", "4")
-        );
-
         logger.info("--> checking index and alias wildcard search");
```
Comment (original-brownbear) [inline server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java:321]: I think it's fine to delete these log lines, there's little to no value in them now that they don't correspond to exactly when the queries run anymore.

**Review by original-brownbear — APPROVED** [review]
LGTM, ideally delete those log lines I pointed out inline before merging though :)

File: server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java (hunk 8 above)
Comment (drempapis) [reply server/src/internalClusterTest/java/org/elasticsearch/aliases/IndexAliasesIT.java:321]: Thank you @original-brownbear. I will proceed with those deletions before merging. I am on the same page

**Pull request merged**